modules/patching/diff.go

```diff
@@ -0,0 +1,275 @@
+package patching
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Diff represents a single change in the document.
+type Diff struct {
+	Insertion  bool
+	StartIndex int
+	Changes    string
+}
+
+// NewDiff creates a new diff object from the given parameters
+func NewDiff(insertion bool, startIndex int, changes string) *Diff {
+	return &Diff{
+		Insertion:  insertion,
+		StartIndex: startIndex,
+		Changes:    changes,
+	}
+}
+
+// NewDiffFromString parses a diff from its string representation.
+func NewDiffFromString(str string) (*Diff, error) {
+	regex, err := regexp.Compile("\\d+:(\\+|-)\\d+:.+")
+	if err != nil {
+		return nil, err
+	}
+	if !regex.MatchString(str) {
+		return nil, errors.New("Illegal patch format; should be %d:+%d:%s or %d:-%d:%s")
+	}
+
+	parts := strings.Split(str, ":")
+	diff := Diff{}
+
+	// Parse startIndex
+	diff.StartIndex, err = strconv.Atoi(parts[0])
+	if err != nil {
+		return nil, fmt.Errorf("Invalid offset: %s", parts[0])
+	}
+
+	// Switch on diff type
+	switch parts[1][0] {
+	case '+':
+		diff.Insertion = true
+	case '-':
+		diff.Insertion = false
+	default:
+		return nil, fmt.Errorf("Invalid operation: %s", parts[1][0:1])
+	}
+
+	// Get length of diff
+	length, err := strconv.Atoi(parts[1][1:])
+	if err != nil {
+		return nil, fmt.Errorf("Invalid length: %s", parts[1][1:])
+	}
+
+	// Un-URL encode the body
+	unescapedChanges, err := url.QueryUnescape(parts[2])
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate changes
+	if length != len(unescapedChanges) {
+		return nil, fmt.Errorf("Length does not match length of change: %d != %s", length, parts[2])
```

Comment:
typically you don't need to have the attribute values that caused an error in the error, that information is valuable in testing, but the error in the actual program should just alert you that the fields were wrong, not what they were.